packages/runed/src/lib/utilities/MediaQuery/MediaQuery.svelte.ts

```diff
@@ -34,43 +35,38 @@ import { browser } from "$lib/internal/utils/browser.js";
  * ```
  */
 export class MediaQuery {
-	#matches: boolean | undefined = $state();
+	#propQuery: MaybeGetter<string>;
+	#query = $derived.by(() => extract(this.#propQuery));
+	#mediaQueryList: MediaQueryList = $derived(window.matchMedia(this.#query));
 	#effectRegistered = false;
-	#query_fn: Getter<string | void> = () => {};
-	// this will be initialized in the constructor
-	#query = $derived(this.#query_fn?.()) as string;
+	#matches: boolean | undefined = $state();
 
 	constructor(query: MaybeGetter<string>) {
-		this.#query_fn = typeof query === "function" ? query : () => query;
-	}
-
-	#matchMedia() {
-		const result = window.matchMedia(this.#query);
-		this.#matches = result.matches;
-		return result;
+		this.#propQuery = query;
 	}
 
 	get matches(): boolean | undefined {
-		// if we are in an effect and this effect has not been registered yet
-		// we match the current value, register the listener and return match
 		if ($effect.active() && !this.#effectRegistered) {
+			this.#matches = this.#mediaQueryList.matches;
+
+			// If we are in an effect and this effect has not been registered yet
+			// we match the current value, register the listener and return match
 			$effect(() => {
-				const result = this.#matchMedia();
 				this.#effectRegistered = true;
-				useEventListener(result, "change", (changed) => {
-					this.#matches = changed.matches;
-				});
-				return () => {
-					this.#effectRegistered = false;
-				};
+
+				useEventListener(
+					() => this.#mediaQueryList,
+					"change",
+					(changed) => (this.#matches = changed.matches)
+				);
+
+				return () => (this.#effectRegistered = false);
 			});
-		} else if (!$effect.active()) {
-			// otherwise if we are not in an effect and the effect has not
-			//been registered we just match media to get the current value
-			if (browser) {
-				this.#matchMedia();
-			}
+		} else if (!$effect.active() && browser) {
+			// Otherwise, just match media to get the current value
+			this.#matches = this.#mediaQueryList.matches;
```

Comment:
I think this could lead to stale values if:

1. i read the value outside an effect
2. the media changes
3. i read the value again outside an effect

Derived are lazily evaluated but are not re-evaluated every time you access them after they have a value.

Reply:
It's not stale because this runs every time you access `current` without an effect running, so it gets revalidated regardless.

Reply:
> It's not stale because this runs every time you access current without an effect running, so it gets revalidated regardless.

Yeah but you are not rerunning matchMedia here, you are just reading the old value no?

Reply:
Ok i was testing and i think `matches` is a getter...it works even without runes lol scratch this comment then